core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -200,21 +203,21 @@ protected void add(DataFile file) {
    * Add a delete file to the new snapshot.
    */
   protected void add(DeleteFile file) {
-    setWriteSpec(file);
-    addedFilesSummary.addedFile(writeSpec(), file);
+    Preconditions.checkNotNull(file, "Invalid delete file: null");
+    PartitionSpec fileSpec = ops.current().spec(file.specId());
+    List<DeleteFile> deleteFiles = newDeleteFiles.computeIfAbsent(file.specId(), specId -> Lists.newArrayList());
+    deleteFiles.add(file);
+    addedFilesSummary.addedFile(fileSpec, file);
```

Comment:
The file spec is only used for partition summaries. I added a test that shows it works as expected.